### PR TITLE
mint-auditor: matching of MobileCoin burns and Gnosis withdrawals

### PR DESCRIPTION
**Pull request by eranrund**
The burn/withdrawal equivalent of #2232

Copy/pasted+minor edits description from #2232:

### Motivation

One of the purposes of the mint auditor is to monitor the MobileCoin blockchain and Gnosis safe(s), in parallel, trying to find matches (or mismatches) between burns (to the MC chain) and withdrawals (from a Gnosis safe).
This PR implements that functionality.

A thing that might be helpful to remember when reviewing this PR is that the two chains sync independently. While the correct ordering for burns/withdrawals is that a burn first takes place on the MobileCoin blockchain, resulting in a burn TxOut added to the ledger and following that, a withdrawal transaction is submitted to the Ethereum blockchain that moves tokens out of the Gnosis safe (while also recoding the TxOut public key of the associated MC burn) it is possible the auditor will first notice the withdrawals and only later learn about the burn TxOut since the syncing of the two separate sources is inherently racey. As such, we need to attempt to find matches both when we learn about a new burn TxOut and when we learn about a new Gnosis withdrawal.

### Future Work
- Expose the match/mismatch data via the GRPC API

### CODE REVIEW

File: mint-auditor/migrations/2022-05-24-214504_create_initial/up.sql [hunk 1]
```diff
@@ -139,6 +156,24 @@ CREATE TABLE audited_mints (
...
 
+-- Mapping between BurnTxOuts and GnosisSafeWithdrawals that match eachother.
+-- This essentially is the audit log that shows which burns/withdrawals were a match.
+-- A match means that the TxOut public key matched a Gnosis safe withdrawal, and that the token information
+-- (MC token_id and Ethereum contract address) as well as the amount all matched.
+-- It is possible for a burn to not be referenced by this table if the burn is not actually associated
+-- with a withdrawal. This is possible since anyone can issue burn transactions. However, a Gnosis withdrawal
+-- is expected to be matched with a burn.
+CREATE TABLE audited_burns (
+   id INTEGER PRIMARY KEY,
```
Comment (remoun) [inline mint-auditor/migrations/2022-05-24-214504_create_initial/up.sql:167]: Nit: indent
```suggestion
    id INTEGER PRIMARY KEY,
```

**Review by remoun — APPROVED** [review]

File: mint-auditor/src/bin/main.rs [hunk 2]
```diff
@@ -342,6 +344,31 @@ fn audit_block_data(
         };
     }
 
+    // Audit burns.
+    for burn_tx_out in &sync_block_data.burn_tx_outs {
+        match AuditedBurn::attempt_match_burn_with_withdrawal(burn_tx_out, config, conn) {
```
Comment (david1alvarez) [inline mint-auditor/src/bin/main.rs:349]: nit: `attempt` -> `try`, matching style on like 325 (unless there is a high-level difference between them that I'm missing)

File: mint-auditor/src/db/schema.rs [hunk 3]
```diff
@@ -2,6 +2,14 @@
 // Run `diesel migration run` to update (inside the mint-auditor directory)
 #![allow(missing_docs)]
 
+table! {
+    audited_burns (id) {
+        id -> Nullable<Integer>,
```
Comment (david1alvarez) [inline mint-auditor/src/db/schema.rs:7]: what do nullable ids get us, compared to non-nullible burn_tx_out_ids?

**Review by david1alvarez — APPROVED** [review]
LGTM

File: mint-auditor/src/db/schema.rs (hunk 3 above)
Comment (eranrund) [reply mint-auditor/src/db/schema.rs:7]: Thats a requirement for SQLite to have auto-increment keys. See the top comment in https://github.com/mobilecoinfoundation/mobilecoin/blob/master/mint-auditor/migrations/2022-05-24-214504_create_initial/up.sql

Comment (eranrund) [reply mint-auditor/src/db/schema.rs:7]: (summary is it has to be nullable to make SQLite happy)

**Pull request merged**